controllers/machinepool_controller.go

```diff
@@ -57,13 +75,179 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
 	r.config = mgr.GetConfig()
-
 	r.scheme = mgr.GetScheme()
 	return nil
 }
 
-func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO(juan-lee): Add machine pool implementation.
-	// TODO(vincepri): Add support for pause annotation and honoring Cluster.Spec.Paused.
+func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+	logger := r.Log.WithValues("machinepool", req.NamespacedName)
+
+	mp := &clusterv1.MachinePool{}
+	if err := r.Client.Get(ctx, req.NamespacedName, mp); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		logger.V(3).Info("Error reading the object - requeue the request.")
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
+	if err != nil {
+		logger.V(3).Info("Failed to get cluster for machinepool.")
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get cluster %q for machinepool %q in namespace %q",
+			mp.Spec.ClusterName, mp.Name, mp.Namespace)
+	}
+
+	// Return early if the object or Cluster is paused.
+	if util.IsPaused(cluster, mp) {
+		logger.V(3).Info("reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(mp, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		r.reconcilePhase(mp)
+		// TODO(jpang): add support for metrics.
+
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, mp); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Reconcile labels.
+	if mp.Labels == nil {
+		mp.Labels = make(map[string]string)
+	}
+	mp.Labels[clusterv1.ClusterLabelName] = mp.Spec.ClusterName
+
+	// Handle deletion reconciliation loop.
+	if !mp.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cluster, mp)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, cluster, mp)
+}
+
+func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	logger := r.Log.WithValues("machinepool", mp.Name, "namespace", mp.Namespace)
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	// If the MachinePool belongs to a cluster, add an owner reference.
+	if cluster != nil && r.shouldAdopt(mp) {
+		mp.OwnerReferences = util.EnsureOwnerRef(mp.OwnerReferences, metav1.OwnerReference{
+			APIVersion: cluster.APIVersion,
+			Kind:       cluster.Kind,
+			Name:       cluster.Name,
+			UID:        cluster.UID,
+		})
+	}
+
+	// If the MachinePool doesn't have a finalizer, add one.
+	if !util.Contains(mp.Finalizers, clusterv1.MachinePoolFinalizer) {
```

Comment:
I wonder if we should start moving to use `AddFinalizerWithError` and `RemoveFinalizerWithError` from [sigs.k8s.io/controller-runtime/pkg/controller/controllerutil](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil) for adding/removing finalizers?

Reply:
Yeah, new issue please. And I think we can use the non-error version?

Reply:
A PR for the filed issue was already merged: https://github.com/kubernetes-sigs/cluster-api/pull/2205, so we should probably adopt that approach for this PR now as well.